### PR TITLE
Patches for Monaco editor

### PR DESCRIPTION
**Pull request by Steve-Mcl**
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

Draft PR with fixes (as found) for monaco editor as [discussed here](https://discourse.nodered.org/t/node-red-2-0-0-beta-1-released/46990/20?u=steve-mcl)

 - Allow setting of Readonly.
 - Make node.status use choices for colour and shape.
 - Let status node also be just a string.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [ ] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js [hunk 1]
```diff
@@ -563,7 +563,7 @@ RED.editor.codeEditor.monaco = (function() {
                     createMonacoCompletionItem("node.send", 'node.send(${1:msg});','async send a msg to the next node',range),
                     createMonacoCompletionItem("node.send (multiple)", 'var ${1:msg1} = {payload:${2:1}};\nvar ${3:msg2} = {payload:${4:2}};\nnode.send([[${1:msg1}, ${3:msg2}]]);','send 1 or more messages out of 1 output',range),
                     createMonacoCompletionItem("node.send (multiple outputs)", 'var ${1:msg1} = {payload:${2:1}};\nvar ${3:msg2} = {payload:${4:2}};\nnode.send([${1:msg1}, ${3:msg2}]);','send more than 1 message out of multiple outputs',range),
-                    createMonacoCompletionItem("node.status", 'node.status({fill:"${1:red}",shape:"${2:ring}",text:"${3:error}"});','Set the status icon and text underneath the function node',range),
+                    createMonacoCompletionItem("node.status", 'node.status({fill:"${1|red,green,yellow,blue,grey|}",shape:"${2|ring,dot|}",text:"${3:message}"});','Set the status icon and text underneath the function node',range),
```
Comment (Steve-Mcl) [inline packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js:566]: As suggested by Dave C - added "node.status snippet enums for fill and shape" 7d24e5b ...
MqPjeAEFE0

**Comment by Steve-Mcl** [comment]
@knolleary Nick, monaco has been bumped to V0.25.2 (from 0.24.0 that we are currently rocking)

I have updated locally and been dogfooding it for a few days.

Should I push this? (to draft or separate PR?)

File: packages/node_modules/@node-red/editor-client/src/types/node-red/func.d.ts [hunk 2]
```diff
@@ -27,56 +27,41 @@ interface NodeStatus {
...
-    text?: string
+    text?: string|boolean|number
 }
 
 declare class node {
     /** 
     * Send 1 or more messages asynchronously 
     * @param {object | object[]} msg  The msg object 
     * @param {Boolean} [clone=true]  Flag to indicate the `msg` should be cloned. Default = `true`   
-    * @example Send 1 msg to output 1
```
Comment (Steve-Mcl) [inline packages/node_modules/@node-red/editor-client/src/types/node-red/func.d.ts:38]: Monaco having problems rendering the example in the intellisense info tip - removed the `@example`

File: packages/node_modules/@node-red/editor-client/src/types/node-red/func.d.ts [hunk 3]
```diff
@@ -27,56 +27,41 @@ interface NodeStatus {
     /** The shape property can be: ring or dot */
     shape?: string,
     /** The text to display */
-    text?: string
+    text?: string|boolean|number
```
Comment (Steve-Mcl) [inline packages/node_modules/@node-red/editor-client/src/types/node-red/func.d.ts:30]: Allow node.status("string only"); 
Reference: https://discourse.nodered.org/t/monaco-clarification/47004/12?u=steve-mcl

File: packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js [hunk 4]
```diff
@@ -667,6 +667,7 @@ RED.editor.codeEditor.monaco = (function() {
                         2307,  //Cannot find module 'xxx' or its corresponding type declarations
                         2322,  //Type 'unknown' is not assignable to type 'string'
                         2339,  //property does not exist on
+                        2345,  //Argument of type xxx is not assignable to parameter of type 'DateTimeFormatOptions'
```
Comment (Steve-Mcl) [inline packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js:673]: Reference: https://discourse.nodered.org/t/monaco-clarification/47004/30?u=steve-mcl

File: packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js [hunk 5]
```diff
@@ -648,6 +648,7 @@ RED.editor.codeEditor.monaco = (function() {
                     strictPropertyInitialization: true,
                     strictFunctionTypes: true,
                     strictBindCallApply: true,
+                    useDefineForClassFields: true,//permit class static fields with private name to have initializer
```
Comment (Steve-Mcl) [inline packages/node_modules/@node-red/editor-client/src/js/ui/editors/code-editors/monaco.js:651]: Reference: https://discourse.nodered.org/t/monaco-clarification/47004/35?u=steve-mcl

**Comment by dceejay** [comment]
@Steve-Mcl - one of the signatures for flow and global context get is the async  get("key","store",callback); where callback is like  function(err,val) { blah... }   - not sure how you can add that but... ;-)

**Comment by knolleary** [comment]
@Steve-Mcl yes we should update to the latest. But as with before, that should be in a separate PR to this one.

In terms of outstanding items, what is remaining? There is the item @dceejay mentions above - the context get/set functions can accept an optional last argument that is a callback - allowing them to be used in async mode. See https://nodered.org/docs/user-guide/writing-functions#asynchronous-context-access for details.

If we can get that fixed in this PR, then mark it ready to review and we'll get this merged.

We can then do the upgrade monaco PR.

If any other items come up, we can create a new PR to collect them in.

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl - one of the signatures for flow and global context get is the async  get("key","store",callback); where callback is like  function(err,val) { blah... }   - not sure how you can add that but... ;-)
> 

No bother, easy enough to sort out. Will do this tomorrow. 

@knolleary apart from Dave's finding, this is all the issues we unearthed in the beta.  I'll fix the issue Dave mentions tomorrow, mark it for review and raise a separate PR for Monaco update

**Comment by Steve-Mcl** [comment]
@dceejay 

>one of the signatures for flow and global context get is the async get("key","store",callback); where callback is like function(err,val) { blah... } - not sure how you can add that but... ;-)

The serverside function node intellisense is built from the JSDoc and typescript in `func.d.ts`

Typescript permits overloads e.g.

```js
    /**
     * Get one or multiple values from context (synchronous).
     * @param name - Name of context variable
     */
    static get(name: string | string[]);

   //below is referring to the same function but is overloaded

    /**
     * Get one or multiple values from context (asynchronous).
     * @param name - Name (or array of names) to get from context
     * @param {function} callback - (optional) Callback function (`(err,value) => {}`)
     */
    static get(name: string | string[], callback: Function);

```

I have now added the overloads - see 3f27dc8 for more.

**Pull request merged**